pkg/controller/operators/openshift/clusteroperator_controller.go

```diff
@@ -0,0 +1,247 @@
+			stripObject(r.co.ClusterOperator)
+			err = r.Client.Create(ctx, r.co.ClusterOperator)
+		}
+
+		// Transient error or successful creation, requeue
+		return r.delayRequeue, err
+	}
+
+	r.co.ClusterOperator = in.DeepCopy()
+
+	var errs []error
+	res := reconcile.Result{}
+	if err := r.mutate(ctx, r.co); err != nil {
+		// Transitive error, requeue
+		log.Error(err, "Error mutating ClusterOperator")
+		errs = append(errs, err)
+		res = r.delayRequeue
+	}
+
+	if !reflect.DeepEqual(r.co.Status, in.Status) {
+		// Status change detected, update
+		if err := r.Client.Status().Update(ctx, r.co.ClusterOperator); err != nil {
+			// Transitive error, requeue
+			errs = append(errs, err)
+			res = r.delayRequeue
+		}
+	}
+
+	return res, utilerrors.NewAggregate(errs)
+}
+
+func (r *ClusterOperatorReconciler) setVersions(_ context.Context, co *ClusterOperator) error {
+	// If we've successfully synced, we know our operator is working properly, so we can update the version
+	if r.syncTracker.SuccessfulSyncs() > 0 && !reflect.DeepEqual(co.Status.Versions, r.TargetVersions) {
+		co.Status.Versions = r.TargetVersions
+	}
+
+	return nil
+}
+
+func (r *ClusterOperatorReconciler) setProgressing(_ context.Context, co *ClusterOperator) error {
+	desired := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorProgressing,
+		LastTransitionTime: r.Now(),
+	}
+
+	if r.syncTracker.SuccessfulSyncs() > 0 && reflect.DeepEqual(co.Status.Versions, r.TargetVersions) {
+		desired.Status = configv1.ConditionFalse
+		desired.Message = fmt.Sprintf("Deployed %s", olmversion.OLMVersion)
+	} else {
+		desired.Status = configv1.ConditionTrue
+		desired.Message = fmt.Sprintf("Waiting to see update %s succeed", olmversion.OLMVersion)
+	}
+
+	current := co.GetCondition(configv1.OperatorProgressing)
+	if conditionsEqual(current, desired) { // Comparison ignores lastUpdated
+		return nil
+	}
+
+	co.SetCondition(desired)
+
+	return nil
+}
+
+func (r *ClusterOperatorReconciler) setAvailable(_ context.Context, co *ClusterOperator) error {
+	desired := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorAvailable,
+		LastTransitionTime: r.Now(),
+	}
+
+	if r.syncTracker.SuccessfulSyncs() > 0 && reflect.DeepEqual(co.Status.Versions, r.TargetVersions) {
+		desired.Status = configv1.ConditionTrue
+	} else {
+		desired.Status = configv1.ConditionFalse
+	}
+
+	current := co.GetCondition(configv1.OperatorAvailable)
+	if conditionsEqual(current, desired) { // Comparison ignores lastUpdated
+		return nil
+	}
+
+	co.SetCondition(desired)
+
+	return nil
+}
+
+func (r *ClusterOperatorReconciler) setDegraded(_ context.Context, co *ClusterOperator) error {
+	desired := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorDegraded,
+		LastTransitionTime: r.Now(),
+	}
+
+	if r.syncTracker.SuccessfulSyncs() > 0 && reflect.DeepEqual(co.Status.Versions, r.TargetVersions) {
```

Comment:
This seems fine, but do we need to lean on reflection? Since OperandVersion is `struct{Name,Version string}` it would be a tiny loop.

Reply:
we can do an `ElementsMatch` type thing easily enough.

Reply:
I went ahead and made this change.